pages/quests/index.tsx

```diff
@@ -8,24 +8,115 @@ import {
+import { useIpfsSubmit, KamonToken } from '@/hooks/useIpfsSubmit'
+import { useToast } from '@/hooks/useToast'
+import { useIpfsGet } from '@/hooks/useIpfsGet'
 
 const Quests: NextPage = () => {
   const { t } = useTranslation('common')
+  const { toast } = useToast()
   const mounted = useMounted()
   const { connect, connectors } = useConnect()
   const [metaMask] = connectors
   const { data } = useAccount()
-  const { keyword, inputChange, submit, isSubmitting } = useKeywordSubmit()
+  const { keyword, inputChange, submit, isSubmitting, keywordSubmitSucceeded } = useKeywordSubmit()
   const { hasNFT } = useHasNFT()
+  const { activeChain } = useNetwork()
+  const kamonNFT = getContractAddress({
+    name: 'kamonNFT',
+    chainId: activeChain?.id
+  })
+  const { balanceOf } = useBalanceOf(kamonNFT, data?.address)
+  const { tokenIdOf } = useTokenIdOf(kamonNFT, data?.address)
+  const { tokenURI } = useTokenURI(kamonNFT, tokenIdOf?.toNumber() || 0)
+  const [tokenId, setTokenId] = useState<BigInt>(BigInt(0))
+  const [tokenJSON, setTokenJSON] = useState<KamonToken>()
+  const [finalTokenUri, setFinalTokenUri] = useState('')
+  const [newTokenImageURI, setNewTokenImageURI] = useState('')
+  const [newTokenJSON, setNewTokenJSON] = useState<KamonToken>()
+  const [updateTxFailed, setUpdateTxFailed] = useState<boolean>()
+
+  const { ipfsSubmit, ipfsSubmitSucceeded, ipfsSubmitIsSubmitting } = useIpfsSubmit()
+  const { ipfsGet, ipfsGetIsSubmitting } = useIpfsGet()
+  const { updateToken, updateTokenSucceeded, updateTokenIsSubmitting } = useUpdateToken(
+    kamonNFT,
+    tokenId,
+    finalTokenUri,
+  )
+  const submittingSomething = () => {
+    return isSubmitting || ipfsSubmitIsSubmitting || ipfsGetIsSubmitting || updateTokenIsSubmitting
+  }
+  useEffect(() => {
+    if (balanceOf && tokenIdOf && tokenURI) {
+      const fetchData = async () => {
+        const pinataRequest = await fetch(tokenURI.toString())
+        const responseJson = await pinataRequest.json()
+        setTokenJSON(responseJson)
+      }
+      fetchData()
+    }
+  }, [balanceOf, tokenIdOf, tokenURI])
+
+  useEffect(() => {
+    if (keywordSubmitSucceeded && !ipfsSubmitIsSubmitting
+      && ipfsSubmitSucceeded == undefined
+      && tokenJSON !== undefined && data?.address !== undefined
+    ) {
+      const userAddress = data?.address
+      const ipfsSubmitWrapper = async () => {
+        const ipfsSubmitRet = await ipfsSubmit(tokenJSON, userAddress)
+        if (ipfsSubmitRet !== 'error' && ipfsSubmitRet !== 'same_points') {
+          const tempFinalTokenUri = await ipfsSubmitRet
+          setFinalTokenUri(tempFinalTokenUri)
+        }
+      }
+      setUpdateTxFailed(false)
+      ipfsSubmitWrapper()
+    }
+  }, [keywordSubmitSucceeded, data?.address, tokenJSON, ipfsSubmit, ipfsSubmitIsSubmitting, ipfsSubmitSucceeded])
+
+  useEffect(() => {
+    if (finalTokenUri && !ipfsGetIsSubmitting && newTokenJSON == undefined) {
```

Comment:
do you need all of the condition ?

Reply:
I reviewed the conditions for all hooks and was able to simplify most of them.
But for this one we might need all three of these conditions:

- `finalTokenUri`: requirement for the code in the hook to run
- `!ipfsGetIsSubmitting`: stop if already running
- `newTokenJSON == undefined`: stop if already completed once